Improves logging and error handling in filesessions.

klizhentas opened the pull request:
Adds chaos tests for async file upload.

klizhentas committed:
Improves logging and error handling in filesessions.

Adds chaos tests for async file upload.

Fixes playback for legacy sessions stored unpacked on disk

webvictim commented on lib/events/filesessions/fileasync_chaos_test.go:
```diff
@@ -0,0 +1,229 @@
+				// for the first 5 resume attempts, simulate nework failure
+				return nil, trace.ConnectionProblem(nil, "failed to resume stream")
+			} else if resumed > 5 && resumed < 8 {
+				// for the next several resumes, loose checkpoint file for the stream
```
```suggestion
				// for the next several resumes, lose checkpoint file for the stream
```

webvictim commented on lib/events/filesessions/fileasync_test.go:
```diff
@@ -359,6 +359,7 @@ func TestUploadResume(t *testing.T) {
 			runResume(t, tc)
 		})
 	}
+
```
```suggestion
```

awly commented on lib/utils/fs_unix.go:
```diff
@@ -56,6 +56,18 @@ func FSReadLock(f *os.File) error {
 
+// FSTryReadLock tries to acquire Flock-style filesystem lock on an open file
+// in read (shared) mode. It returns CompareFailed if the lock is already acquired.
+func FSTryReadLock(f *os.File) error {
```
This seems to be unused

fspmarshall reviewed (approved):
LGTM once @awly's suggestions are implemented.

klizhentas commented on lib/events/filesessions/fileasync_chaos_test.go:
```diff
@@ -0,0 +1,229 @@
+// TestUploadChaos introduces failures in all stages of the async
+// upload process and verifies that the system is working correctly.
+//
+// Data race detector slows down the test significantly (10x+),
+// that is why the test is skipped when tests are running with
+// `go test -race` flag or `go test -short` flag
```
We run integration test with race detector as well. I think we should have a separate class of tests we run without race detector.

I would like to have tests that run for a while and try to break the system in different ways.

klizhentas replied:
@webvictim can we add a line to run `go test -test.run=TestChaos ./lib/....` to activate those tests?

klizhentas replied:
I stand corrected, integration tests are running without `-race` flag, however I think it will be good to have `TestChaos` step that is local, but does not activate `-race` detector.

klizhentas replied:
@webvictim @awly I've added another build target `make chaos` and `make -C build.assets chaos` alongside extra drone step. What do you folks think? Take a look and let me know please.

awly replied:
For others' reference, `TestChaosUpload` takes:
- 0.3s without race detector
- 12.4s with race detector
12s is slow enough to be annoying, although we have much slower unit tests.

I'd prefer to either add it to `make test` without race detector or `make integration` with (eventually).

If we're going to add a new category of tests, it could be based on how they are used:
- `make test` - "fast" tests for local dev iteration
- `make integration` - "slow" tests for each PR running in CI
- `make stress` - "infinite" tests that hammer at the code for a fixed period of time, running as a cronjob
  - fuzzers also fit here

P.S. integration tests don't run with the race detector because we didn't get to fixing all the races yet. They should eventually run with the race detector.

webvictim replied:
The biggest problem with adding this as a completely separate step to Drone is the fact that the tests take 0.3 seconds to run but the setup (downloading the buildbox image, starting the Docker container etc) takes 22 seconds. Given that our unit test runs are already taking ~4-5 minutes and integration tests ~4-5 minutes, I don't think it's a huge problem to add them to one of the other runs.

I agree with @awly's idea of saying that ultimately, all unit tests should be 'quick' tests you can run locally for iterative development. Anything that takes a long time should be part of integration.

klizhentas commented on lib/events/filesessions/fileasync_chaos_test.go:
```diff
@@ -0,0 +1,229 @@
+		t.Skip("Skipping chaos test in short mode.")
+	}
+
+	utils.InitLoggerForTests(testing.Verbose())
```
I think`InitLoggerForTests` should modify global logger in tests to use `FailureOnly` handler.

klizhentas replied:
Because right now `FailureOnly` has to be passed manually everywhere

awly replied:
Modifying the global logger messes up tests running in parallel.
We already pass logger objects via most `*Config` structs, let's stick with that pattern?

klizhentas commented:
@awly @fspmarshall please take a look again folks, I've added `Chaos` tests to run without data race here. I've also found another unrelated bug with playback with legacy sessions and added it as a separate commit to the same branch.

fspmarshall commented on Makefile:
```diff
@@ -266,6 +266,15 @@ integration:
+#
+.PHONY: chaos
+chaos: FLAGS ?= -v -cover
+chaos: FOLDERS := $(shell find . -type f -name '*chaos*' | xargs dirname | uniq)
```
This pulls in matches from `vendor` and non-go directories.  Probably more future-proof to do something like this:

```suggestion
chaos: FOLDERS := $(shell find . -type f -name '*chaos*.go' -not -path "*/vendor/*" | xargs dirname | uniq)
```

klizhentas commented:
@awly @webvictim @fspmarshall take another look please:

* I did not change the logging in tests for now, consider it a minor future improvement
* Added `chaos` tests to run as the same `test` run as recommended by Gus
* Improved run to exclude potential matches in vendor

klizhentas commented:
@awly also made `make test` to run all tests except integration.

webvictim reviewed (approved):
LGTM on the CI parts

awly commented on lib/events/auditlog_test.go:
```diff
@@ -392,6 +392,65 @@ func (a *AuditTestSuite) TestForwardAndUpload(c *check.C) {
 
+// TestLegacyHandler tests playback for legacy sessions
+// that are stored on disk in unpacked format
+func (a *AuditTestSuite) TestLegacyHandler(c *check.C) {
```
optional: convert this to the new test style

klizhentas commented:
retest this

webvictim commented:
retest this **please**

klizhentas commented:
retest this please
